app/src/main/java/de/westnordost/streetcomplete/screens/about/LogsFragment.kt

```diff
@@ -0,0 +1,145 @@
+        binding.logsList.itemAnimator = null // default animations are too slow when logging many messages quickly
+        binding.logsList.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
+
+        if (savedInstanceState != null) {
+            onLoadInstanceState(savedInstanceState)
+        }
+
+        showLogs()
+
+        logsController.addListener(logsControllerListener)
+    }
+
+    private fun onLoadInstanceState(savedInstanceState: Bundle) {
+        filters = Json.decodeFromString(savedInstanceState.getString(FILTERS_DATA)!!)
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putString(FILTERS_DATA, Json.encodeToString(filters))
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        logsController.removeListener(logsControllerListener)
+    }
+
+    private fun createOptionsMenu(toolbar: Toolbar) {
+        toolbar.inflateMenu(R.menu.menu_logs)
+
+        toolbar.setOnMenuItemClickListener { item ->
+            when (item.itemId) {
+                R.id.action_share -> {
+                    onClickShare()
+                    true
+                }
+                R.id.action_filter -> {
+                    onClickFilter()
+                    true
+                }
+                else -> false
+            }
+        }
+    }
+
+    private fun onClickShare() = viewLifecycleScope.launch {
+        val shareIntent = Intent(Intent.ACTION_SEND).apply {
+            val logText = getLogs().format()
+            val logTimestamp = LocalDateTime.now().toString()
+            val logTitle =
+                "${BuildConfig.APPLICATION_ID}_${BuildConfig.VERSION_NAME}_$logTimestamp.log"
+
+            putExtra(Intent.EXTRA_TEXT, logText)
+            putExtra(Intent.EXTRA_TITLE, logTitle)
+            type = "text/plain"
+        }
+
+        startActivity(Intent.createChooser(shareIntent, null))
+    }
+
+    private fun onClickFilter() {
+        LogsFiltersDialog(requireContext(), filters) { newFilters ->
+            filters = newFilters
+            showLogs()
+        }.show()
+    }
+
+    private fun onMessageAdded(message: LogMessage) {
+        if (filters.matches(message)) {
+            adapter.add(message)
+            binding.toolbar.root.title = getString(R.string.about_title_logs, adapter.messages.size)
+
+            if (hasScrolledToBottom()) {
+                binding.logsList.scrollToPosition(adapter.messages.lastIndex)
+            }
+        }
+    }
+
+    private fun showLogs() {
+        viewLifecycleScope.launch {
+            val logs = getLogs()
+            adapter.messages = logs
+            binding.toolbar.root.title = getString(R.string.about_title_logs, logs.size)
+            binding.logsList.scrollToPosition(logs.lastIndex)
+        }
+    }
+
+    private suspend fun getLogs(): List<LogMessage> = withContext(Dispatchers.IO) {
```

Comment:
By the way, this means that on opening the log view, in absence of any filters, it will load *all logs* from the database, i.e. logs spanning over several days. The index doesn't help here.

A workaround might be to add some sort of internal pagination that uses `LIMIT $limit` for the first query and on scrolling, for `newerThan`/`olderThan` is passed to the DAO equal to the first/last item in the list. (However, this solution would have the caveat, that the scroll bar always has the wrong size)

But this seems to be a bit of work and adds some complexity. And if/when the UI code is changed to Jetpack Compose / Multiplatform Compose, the implementation would need to look different anyway. Maybe there is even a facilitation for this, as this seems to be a common use case. So in any case, as long as this does not turn to be a huge problem, I think it is fine to leave it as it is now, without any sort of internal pagination/buffering.

Reply:
Actually, a really easy workaround came to my mind: The default filters when entering the fragment are only the logs from today.

```kotlin
    init {
        val startOfToday = LocalDateTime(systemTimeNow().toLocalDate(), LocalTime(0, 0, 0))
        filters = LogsFilters(timestampNewerThan = startOfToday)
    }
```